airflow/dag_processing/manager.py

```diff
@@ -576,8 +585,11 @@ def _run_parsing_loop(self):
         while True:
             loop_start_time = time.monotonic()
             ready = multiprocessing.connection.wait(self.waitables.keys(), timeout=poll_time)
-            if self._job:
-                self._job.heartbeat()
+            # we cannot (for now) define job in _job_runner nicely due to circular references of
+            # job and job runner, so we have to use hasattr, but we might address it in the future
+            # change when decoupling these two even more
+            if hasattr(self._job_runner, "job") and self._job_runner.job is not None:
```

Comment:
```suggestion
            if getattr(self._job_runner, "job", None) is not None:
```

Reply:
BTW. I removed this anyway in the final refactoring step: https://github.com/apache/airflow/pull/30376 (which I simplified quite a bit)